File: src/main/java/wraith/fabricaeexnihilo/modules/tools/ToolItemWithRegistry.java
package wraith.fabricaeexnihilo.modules.tools;

import net.fabricmc.fabric.api.item.v1.FabricItemSettings;
import net.minecraft.block.BlockState;
import net.minecraft.entity.EquipmentSlot;
import net.minecraft.entity.LivingEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ToolItem;
import net.minecraft.item.ToolMaterial;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import wraith.fabricaeexnihilo.api.registry.ToolRecipeRegistry;

public abstract class ToolItemWithRegistry extends ToolItem {

    private final ToolRecipeRegistry registry;
    private final float blockBreakingSpeed;

    public ToolItemWithRegistry(ToolMaterial material, ToolRecipeRegistry registry, FabricItemSettings settings) {
        super(material, settings);
        this.registry = registry;
        this.blockBreakingSpeed = material.getMiningSpeedMultiplier();
    }

    @Override
    public boolean isSuitableFor(BlockState state) {
        return registry.isRegistered(state.getBlock());
    }

    @Override
    public float getMiningSpeedMultiplier(ItemStack stack, BlockState state) {
        return isSuitableFor(state) ? blockBreakingSpeed : 1F;
    }

    @Override
    public boolean postMine(ItemStack stack, World world, BlockState state, BlockPos pos, LivingEntity miner) {
        if (!world.isClient && state.getHardness(world, pos) != 0.0f) {
            stack.damage(1, miner, ent -> ent.sendEquipmentBreakStatus(EquipmentSlot.MAINHAND));
        }
        return true;
    }

}

File: src/main/kotlin/exnihilofabrico/common/tools/ToolItemWithRegistry.kt
package exnihilofabrico.common.tools

import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.api.registry.IToolRegistry
import net.minecraft.block.BlockState
import net.minecraft.entity.EquipmentSlot
import net.minecraft.entity.LivingEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.ToolItem
import net.minecraft.item.ToolMaterial
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

abstract class ToolItemWithRegistry(material: ToolMaterial, val registry: IToolRegistry, settings: Item.Settings):
    ToolItem(material, settings.group(ExNihiloFabrico.ITEM_GROUP)) {

    private val blockBreakingSpeed: Float = material.miningSpeed

    override fun isEffectiveOn(state: BlockState) = registry.isRegistered(state.block)

    override fun getMiningSpeed(stack: ItemStack, state: BlockState) =
        if(isEffectiveOn(state)) blockBreakingSpeed else 1f

    override fun postMine(stack: ItemStack, world: World, state: BlockState, pos: BlockPos, entity: LivingEntity?): Boolean {
        if (!world.isClient && state.getHardness(world, pos) != 0.0f)
            stack.damage(1, entity, { ent -> ent?.sendEquipmentBreakStatus(EquipmentSlot.MAINHAND) })
        return true
    }
}